File: Questions/LC Questions/Arrays/2294. Partition Array Such That Maximum Difference Is K/solution.ts
// TC: O(nlogn), SC: O(1)
function partitionArray(nums: number[], k: number): number {
    const n = nums.length
    nums.sort((a, b) => a - b)

    let start_of_valid_subsequence = nums[0]
    let count = 1

    for (let i = 1; i < n; i++) {
        if (nums[i] - start_of_valid_subsequence > k) {
            start_of_valid_subsequence = nums[i]
            count++
        }
    }
    return count
};

File: Questions/LC Questions/Arrays/2294. Partition Array Such That Maximum Difference Is K/solution.py
from typing import List

# TC: O(nlogn), SC: O(1)
class Solution:
    def partitionArray(self, nums: List[int], k: int) -> int:
        n = len(nums)
        nums.sort()
        start_of_valid_subsequence = nums[0]
        count = 1

        for i in range(1, n):
            if nums[i] - start_of_valid_subsequence > k:
                start_of_valid_subsequence = nums[i]
                count += 1
        return count